### PR TITLE
Feature: Portfolio calculation

### PR DESCRIPTION
**Pull request by chumaknadya**

### CODE REVIEW

**Comment by calj** [comment]
- [ ] Raise an exception if there is no trade on a market
- [ ] Raise an exception if a market is missing for straight forward conversion

File: app/jobs/cron/portfolio.rb [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        res.present? ? res : 0
+      end
+
+      def portfolio_currencies
+        ENV.fetch('PORTFOLIO_CURRENCIES', '').split(',')
+      end
+
+      def conversion_market(currency, portfolio_currency)
+        market_name = "#{currency}#{portfolio_currency}"
+        raise Error, "There is no market #{market_name}" unless Market.find_by(id: market_name).present?
```
Comment (calj) [inline app/jobs/cron/portfolio.rb]: we should find the market by base and quote currencies here, the market id isn't normalized.

File: spec/api/v2/account/portfolio_spec.rb [hunk 2]
```diff
@@ -5,73 +5,87 @@
...
-    let!(:btcusd_trade) { create(:trade, :btcusd, price: 8200, amount: 0.1, maker_order: btcusd_ask, taker_order: btcusd_bid) }
+      expect(response).to be_successful
 
-    let(:btcusd_pa) { { 'base_unit' => 'btc', 'price' => '8200.0', 'total' => '820.0' } }
-    let(:ethusd_pa) { { 'base_unit' => 'eth', 'price' => '300.0', 'total' => '360.0' } }
+      expect(response_body.count).to eq 2
+      expect(response_body[0]['currency']).to eq(portfolio1.currency_id)
+      expect(response_body[0]['portfolio_currency']).to eq(portfolio1.portfolio_currency_id)
+      expect(response_body[0]['total_credit'].to_f).to eq(portfolio1.total_credit)
+      expect(response_body[0]['total_credit_fees'].to_f).to eq(portfolio1.total_credit_fees)
```
Comment (calj) [inline spec/api/v2/account/portfolio_spec.rb]: we don't need to expose fees through the user API

File: spec/api/v2/account/portfolio_spec.rb [hunk 3]
```diff
@@ -5,73 +5,87 @@
...
+      expect(response_body[0]['total_debit_fees'].to_f).to eq(portfolio1.total_debit_fees)
+      expect(response_body[0]['total_credit_value'].to_f).to eq(portfolio1.total_credit_value)
+      expect(response_body[0]['total_debit'].to_f).to eq(portfolio1.total_debit)
+      expect(response_body[0]['total_debit_value'].to_f).to eq(portfolio1.total_debit_value)
+      expect(response_body[0]['avarage_buy_price'].to_f.round(9)).to eq( (portfolio1.total_credit_value / (portfolio1.total_credit + portfolio1.total_credit_fees)).to_f)
```
Comment (calj) [inline spec/api/v2/account/portfolio_spec.rb]: typo average

File: app/jobs/cron/portfolio.rb [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    Error = Class.new(StandardError)
+
+    class <<self
+      def max_liability(portfolio_currency)
+        res = ::Portfolio.where(portfolio_currency_id: portfolio_currency).maximum('last_liability_id')
+        res.present? ? res : 0
+      end
+
+      def portfolio_currencies
+        ENV.fetch('PORTFOLIO_CURRENCIES', '').split(',')
```
Comment (mod) [inline app/jobs/cron/portfolio.rb:12]: fuck validations

**Review by mod — APPROVED** [review]
EXPERIMENTAL FEATURE PNL

**Pull request merged**